core/src/heaviest_subtree_fork_choice.rs

```diff
@@ -858,37 +858,32 @@ impl HeaviestSubtreeForkChoice {
         tower
             .last_voted_slot_hash()
             .and_then(|last_voted_slot_hash| {
-                let heaviest_slot_hash_on_same_voted_fork = self.best_slot(&last_voted_slot_hash);
-                if heaviest_slot_hash_on_same_voted_fork.is_none() {
-                    if !tower.is_stray_last_vote() {
-                        // Unless last vote is stray and stale, self.bast_slot(last_voted_slot) must return
-                        // Some(_), justifying to panic! here.
-                        // Also, adjust_lockouts_after_replay() correctly makes last_voted_slot None,
-                        // if all saved votes are ancestors of replayed_root_slot. So this code shouldn't be
-                        // touched in that case as well.
-                        // In other words, except being stray, all other slots have been voted on while this
-                        // validator has been running, so we must be able to fetch best_slots for all of
-                        // them.
-                        panic!(
-                            "a bank at last_voted_slot({:?}) is a frozen bank so must have been \
-                            added to heaviest_subtree_fork_choice at time of freezing",
-                            last_voted_slot_hash,
-                        )
-                    } else {
-                        // fork_infos doesn't have corresponding data for the stale stray last vote,
-                        // meaning some inconsistency between saved tower and ledger.
-                        // (newer snapshot, or only a saved tower is moved over to new setup?)
-                        return None;
+                match self.is_candidate(&last_voted_slot_hash) {
+                    Some(true) => self.best_slot(&last_voted_slot_hash),
+                    Some(false) => None,
```

Comment:
This may be a fine change, but trying to minimize as many logical changes as possible.

In this new code `is_candidate(&last_voted_slot_hash) == false`, this returns None. This seems like it might be a deviation from the original code.

To maintain the same behavior would require that in the new code if `self.is_candidate(&last_voted_slot_hash) == Some(false)` then that implies inthe old code that `heaviest_slot_hash_on_same_voted_fork == None` in the old code, which I don't think is necessarily true?

Reply:
I'm having a hard time understanding what you're suggesting that I change here

Reply:
> To maintain the same behavior would require that in the new code if self.is_candidate(&last_voted_slot_hash) == Some(false) then that implies in the old code that heaviest_slot_hash_on_same_voted_fork == None in the old code, which I don't think is necessarily true?
I think he's saying that this is a new condition that wasn't checked in the old code that might now return None.

Reply:
> To maintain the same behavior would require that in the new code if self.is_candidate(&last_voted_slot_hash) == Some(false) then that implies in the old code that heaviest_slot_hash_on_same_voted_fork == None in the old code, which I don't think is necessarily true?

I think he's saying that this is a new condition that wasn't checked in the old code that might now return None. 
I don't think this implies in the old code that `heaviest_slot_hash_on_same_voted_fork` needed to be None, this just requires that there is an invalid ancestor which wasn't being checked before.

Reply:
agreed that we should try to reduce the change set to the bare minimum here.  we need to fix the known bug _now_, further hardening can wait.

kinda confusing that a comment like this would result in an approved review. seems blocking

Reply:
I think the original code was supposed to be checking whether the slot was in `fork_infos` or not and that's the behavior that I retained. The new behavior is that this function should have been filtering out slots that were invalid rather than doing the last vote equality check. I think both changes here are necessary.

Reply:
The change is already at the bare minimum in my opinion. Take note that `self.is_candidate(..).is_none()` and `self.best_slot(..).is_none()` is the same thing because each of them is only `None` when the queried slot is not in `fork_infos`.

The only logical change is switching from filtering the returned heaviest slot by `fork_info.is_candidate` rather than checking if it is the same slot as the last vote. Checking `is_candidate` is crucial here because otherwise we would build blocks from heaviest bank even if it was a duplicate.

Reply:
@t-nelson @carllin if you disagree on what the minimum change, please let me know. I'm also in favor of keeping the scope of change as low as possible and it's why I refrained from going further and adding a check for `is_candidate` in `best_slot` (which I believe to be a more complete fix)

Reply:
After more digging and discussion with @carllin, it looks like we have to add the `is_candidate` check after all because otherwise we could get into situations where the `heaviest_slot_on_same_voted_fork` was purged (because the cluster confirmed a different version) and we try to reset to a purged bank. I think that avoiding resetting to invalid forks should be avoided at all costs.

Reply:
> I think that avoiding resetting to invalid forks should be avoided at all costs.

this would be separate from the condition that halted MB though, right? i'm not saying that we shouldn't harden further, just that we can run anything extra through the usual (priority) stabilization/backport process

Reply:
We did agree to add the is_candidate check in the normal process but without it we expose validators to increased risk of panicking due to purged slots.